distribution/cachedir/manager.go

```diff
@@ -0,0 +1,190 @@
+	return os.RemoveAll(c.cachePathForKeyHash(keyHash))
+}
+
+func (c *manager) GetDir(key string) string {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	hashedKey := hashKey(key)
+	logrus.WithField("key", key).WithField("hashedKey", hashedKey).Debugf("cachedir manager get")
+	if _, ok := c.inUse[hashedKey]; ok {
+		panic(fmt.Errorf("tried to get a cachedir key (%s) twice without releasing/deleting", key))
+	}
+	c.inUse[hashedKey] = struct{}{}
+
+	return c.cachePathForKeyHash(hashedKey)
+}
+
+func (c *manager) ReleaseDir(key string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	hashedKey := hashKey(key)
+	logrus.WithField("key", key).WithField("hashedKey", hashedKey).Debugf("cachedir manager release")
+	if _, ok := c.inUse[hashedKey]; !ok {
+		panic(fmt.Errorf("tried to release a cachedir key (%s) without getting it first", key))
+	}
+	delete(c.inUse, hashedKey)
+}
+
+func (c *manager) DeleteDir(key string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	hashedKey := hashKey(key)
+	logrus.WithField("key", key).WithField("hashedKey", hashedKey).Debugf("cachedir manager delete")
+	if _, ok := c.inUse[hashedKey]; !ok {
+		panic(fmt.Errorf("tried to delete a cachedir key (%s) without getting it first", key))
+	}
+	delete(c.inUse, hashedKey)
+	return c.cleanUpByKeyHash(hashedKey)
+}
+
+func (c *manager) CollectGarbage() (uint64, error) {
+	// safety check to make sure we don't try to delete files in / or pwd
+	if c.cacheRoot == "" {
+		return 0, nil
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// mark
+	allKeyHashes := map[HashedKey]struct{}{}
+	cacheDirs, err := ioutil.ReadDir(c.cacheRoot)
+	if err != nil {
+		logrus.WithError(err).WithField("cacheRoot", c.cacheRoot).Warnf("Failed to list cache root")
+	}
+	for _, cacheDir := range cacheDirs {
+		cacheDirName := cacheDir.Name()
+		keyHash := HashedKey(filepath.Base(cacheDirName))
+
+		// safety check to make sure we don't delete anything that's not a hash
+		// we created
+		var decoded [sha256.Size]byte
+		l, err := hex.Decode(decoded[:], []byte(keyHash))
+		if err != nil {
+			logrus.WithError(err).WithField("cachedir", cacheDirName).Debug("Ignoring unexpected cache directory")
+			continue
+		}
+		if l != sha256.Size {
+			logrus.WithField("cachedir", cacheDirName).Debug("Ignoring unexpected cache directory because the hash's decoded length is %s, not %s", l, sha256.Size)
+			continue
+		}
+
+		allKeyHashes[keyHash] = struct{}{}
+	}
+
+	// set subtraction allKeyHashes = allKeyHashes - c.inUse
+	for key := range c.inUse {
+		cacheDirName := c.cachePathForKeyHash(key)
+		keyHash := HashedKey(filepath.Base(cacheDirName))
+		delete(allKeyHashes, keyHash)
+	}
+
+	// sweep
+	var total uint64
+	errs := []error{}
+	for keyHash := range allKeyHashes {
+		total += c.sizeEstimateByKeyHash(keyHash)
+		err = c.cleanUpByKeyHash(keyHash)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) > 0 {
+		return total, fmt.Errorf("failed to delete some files during cachedir manager CollectGarbage: %s", errs)
```

Comment:
The formatting of multiple errors here is not ideal: https://play.golang.org/p/RhjEF3WM2W